administrator/components/com_workflow/src/Controller/GraphController.php

```diff
@@ -0,0 +1,307 @@
+                    'published'   => (bool) $stage->published,
+                    'default'     => (bool) $stage->default,
+                    'ordering'    => (int) $stage->ordering,
+                    'position'    => $stage->position ? json_decode($stage->position, true) : null,
+                    'workflow_id' => $stage->workflow_id,
+                ];
+            }
+            echo new JsonResponse($response);
+        } catch (\Exception $e) {
+            echo new JsonResponse($e->getMessage(), 'error', true);
+        }
+
+        $this->app->close();
+    }
+
+
+    /**
+     * Retrieves all transitions for the specified workflow to be used in the workflow graph view.
+     *
+     * Fetches transitions by workflow ID and returns the result as a JSON response
+     * for graphical editors or API consumers.
+     *
+     * @return  void  Outputs a JSON response with transitions data or error message.
+     *
+     * @since   _DEPLOY_VERSION_
+     */
+    public function getTransitions()
+    {
+        try {
+            $workflowId = $this->input->getInt('workflow_id');
+            $model = $this->getModel('Transitions');
+
+            $model->setState('filter.workflow_id', $workflowId);
+            $model->setState('list.limit', 0);
+
+            $transitions = $model->getItems();
+            $response = [];
+
+            foreach ($transitions as $transition) {
+                $response[] = [
+                    'id'            => (int) $transition->id,
+                    'title'         => $transition->title,
+                    'description'   => $transition->description,
+                    'published'     => (bool) $transition->published,
+                    'from_stage_id' => (int) $transition->from_stage_id,
+                    'to_stage_id'   => (int) $transition->to_stage_id,
+                    'ordering'      => (int) $transition->ordering,
+                    'workflow_id'   => (int) $transition->workflow_id
+                ];
+            }
+
+            echo new JsonResponse($response);
+        } catch (\Exception $e) {
+            echo new JsonResponse($e->getMessage(), 'error', true);
+        }
+
+        $this->app->close();
+    }
+
+    /**
+     * Method to save stage positions
+     *
+     * @return  boolean  True if successful, false otherwise.
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    public function save()
+    {
+        // Check for request forgeries
+        $this->checkToken();
+
+        $app = $this->app;
+        $input = $app->input;
+        $workflowId = $input->getInt('id');
+        $extension = $this->input->getCmd('extension');
+        $options = $input->getCmd('option');
+
+        $task = $this->getTask();
+
+        $positionsJson = $input->getString('positions', '');
+        $positions = [];
+
+        if (!empty($positionsJson)) {
+            try {
+                $positions = json_decode($positionsJson, true);
```

Comment:
There should be some kind of validation, that the position has the correct structure.